File: example/ios/Runner/Biz/PowerImageNetworkImageLoader.m
//
//  MyNetworkImageLoader.m
//  Runner
//
//  Created by 王振辉 on 2021/7/15.
//

#import "PowerImageNetworkImageLoader.h"
#import <SDWebImage/SDWebImage.h>
#import "PowerFlutterImage.h"
#import "PowerFlutterMultiFrameImage.h"

@implementation PowerImageNetworkImageLoader
- (void)handleRequest:(PowerImageRequestConfig *)requestConfig completed:(PowerImageLoaderCompletionBlock)completedBlock {
    
    /// CDN optimization, you need transfer reqSize to native image loader!
    /// CDN optimization, you need transfer reqSize to native image loader!
    /// like this: [[SDWebImageManager sharedManager] downloadImageWithURL:[NSURL URLWithString:requestConfig.srcString] viewSize:reqSize completed:
    CGSize reqSize = requestConfig.originSize;
    /// attention.

    
    [[SDWebImageManager sharedManager] loadImageWithURL:[NSURL URLWithString:requestConfig.srcString] options:nil progress:^(NSInteger receivedSize, NSInteger expectedSize, NSURL * _Nullable targetURL) {

        } completed:^(UIImage * _Nullable image, NSData * _Nullable data, NSError * _Nullable error, SDImageCacheType cacheType, BOOL finished, NSURL * _Nullable imageURL) {
            if (image != nil) {
                if (image.sd_isAnimated) {
                    NSArray<SDImageFrame *> *frames = [SDImageCoderHelper framesFromAnimatedImage:image];
                    if (frames.count > 0) {
                        NSMutableArray *array = [NSMutableArray new];
                        for (int i = 0; i < frames.count; i++) {
                            SDImageFrame *frame = frames[i];
                            [array addObject:[PowerImageFrame frameWithImage:frame.image duration:frame.duration]];
                        }

                        PowerFlutterImage *flutterImage = [[PowerFlutterMultiFrameImage alloc] initWithImage:image frames:array];
                        completedBlock([PowerImageResult successWithPowerFlutterImage:flutterImage]);
                        return;
                    }
                }
                completedBlock([PowerImageResult successWithImage:image]);
            }else {
                completedBlock([PowerImageResult failWithMessage:error.localizedDescription]);
            }
    }];
    
}


@end

File: example_kotlin_swift/ios/Runner/Loaders/PowerImageNetworkImageLoader.swift
//
//  PowerImageNetworkImageLoader.swift
//  Runner
//
//  Created by wjr on 2022/5/30.
//

import power_image
import SDWebImage

class PowerImageNetworkImageLoader:NSObject,PowerImageLoaderProtocol{
    
    func handleRequest(_ requestConfig: PowerImageRequestConfig!, completed completedBlock: PowerImageLoaderCompletionBlock!) {
        let reqSize:CGSize = requestConfig.originSize
        let url = URL(string: requestConfig.srcString())
        SDWebImageManager.shared.loadImage(with: url, progress: nil) { image, data, error, cacheType, finished, url in
            
            if let image = image {
                if (image.sd_isAnimated) {
                    let frames:[SDImageFrame] = SDImageCoderHelper.frames(from: image)!
                    if frames.count > 0 {
                        var arr:[PowerImageFrame] = []
                        for index in 0..<frames.count {
                            let frame:SDImageFrame = frames[index]
                            arr.append(PowerImageFrame(image: frame.image, duration: frame.duration))
                        }
                        let flutterImage = PowerFlutterMultiFrameImage(image: image, frames: arr)
                        completedBlock(PowerImageResult.success(with: flutterImage))
                        return
                    }
                }
                
                completedBlock(PowerImageResult.success(with: image))
                
            }else{
                completedBlock(PowerImageResult.fail(withMessage: error?.localizedDescription ?? "PowerImageNetworkLoaderError!"))
            }   
        }
    }
}
